neps/nep-0591.md

```diff
@@ -0,0 +1,224 @@
+
+We introduce new receipt action for deploying global contracts:
+
+```rust
+struct DeployGlobalContractAction {
+    code: Vec<u8>,
+    deploy_mode: GlobalContractDeployMode,
+}
+
+enum GlobalContractDeployMode {
+    /// Contract is deployed under its code hash.
+    /// Users will be able reference it by that hash.
+    /// This effectively makes the contract immutable.
+    CodeHash,
+    /// Contract is deployed under the owner account id.
+    /// Users will be able reference it by that account id.
+    /// This allows the owner to update the contract for all its users.
+    AccountId,
+}
+```
+
+User pays for storage by burning NEAR tokens from its balance depending on the deployed contract size.
+Global contract is not checked to be compilable wasm (just like in case of a regular contract), so it is possible to deploy invalid wasm and that still burns tokens.
+
+Also new action is added for using previously deployed global contract:
+
+```rust
+struct UseGlobalContractAction {
+    contract_identifier: GlobalContractIdentifier,
+}
+
+enum GlobalContractIdentifier {
+    CodeHash(CryptoHash),
+    AccountId(AccountId),
+}
+```
+
+This action is similar to deploying a regular contract to an account, except user does not cover storage deposit.
+Using non-existing global contract (both by its hash and account id) results in `GlobalContractDoesNotExist` action error, so users have to wait for global contract distribution to be completed in order to starting using the contract.
+
+## Reference Implementation
+
+### Storage
+
+In order to have global contracts available to users on all shards we store a copy in each shard's trie.
+A new trie key is introduced for that:
+
+```rust
+pub enum TrieKey {
+    ...
+    GlobalContractCode {
+        identifier: GlobalContractCodeIdentifier,
+    },
+}
+
+pub enum GlobalContractCodeIdentifier {
+    CodeHash(CryptoHash),
+    AccountId(AccountId),
+}
+```
+
+The value is contract code bytes, similar to `TrieKey::ContractCode`.
+
+### Distribution
+
+Global contract has to be distributed to all shards after being deployed.
+This is implemented with a dedicated receipt type:
+
+```rust
+enum ReceiptEnum {
+    ...
+    GlobalContractDistribution(GlobalContractDistributionReceipt),
+}
+
+enum GlobalContractDistributionReceipt {
+    V1(GlobalContractDistributionReceiptV1),
+}
+
+struct GlobalContractDistributionReceiptV1 {
+    id: GlobalContractIdentifier,
+    target_shard: ShardId,
+    already_delivered_shards: Vec<ShardId>,
+    code: Arc<[u8]>,
+}
+```
+
+`GlobalContractDistribution` receipt is generated as a result of processing `DeployGlobalContractAction`.
+Such receipt contains destination shard `target_shard` as well as list of already visited shard ids `already_delivered_shards`.
+Applying `GlobalContractDistribution` receipt updates the corresponding `TrieKey::GlobalContractCode` in the trie for the current shard.
```

Comment:
Just to clarify, this means that a contract becomes available in a shard after processing the `DeployGlobalContractAction` action, so the contract can be used in some shards before other ones right?

Reply:
In case this is correct, some "funny" side effect, is that two contracts living in different shards, pointing at the same global contract, could run an action on the same block, and be using two different contracts, since the update is still in flight.

By "funny" I mean that it is hard to foresee side effects of this behavior, or if it could be exploited.

Reply:
You are right, this can result in such "funny" behaviours. Effectively during upgrade 2 versions of the contract will be active on different shards for some range of blocks. That is something contract developers should be aware of. Bowen also raised that point and we reached a conclusion that it should not be a big issue in practice.